flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaWriter.java

```diff
@@ -0,0 +1,379 @@
+     * @param recordSerializer serialize to transform the incoming records to {@link ProducerRecord}
+     * @param schemaContext context used to initialize the {@link KafkaRecordSerializationSchema}
+     * @param recoveredStates state from an previous execution which was covered
+     */
+    KafkaWriter(
+            DeliveryGuarantee deliveryGuarantee,
+            Properties kafkaProducerConfig,
+            String transactionalIdPrefix,
+            Sink.InitContext sinkInitContext,
+            KafkaRecordSerializationSchema<IN> recordSerializer,
+            SerializationSchema.InitializationContext schemaContext,
+            List<KafkaWriterState> recoveredStates) {
+        this.deliveryGuarantee = checkNotNull(deliveryGuarantee, "deliveryGuarantee");
+        this.kafkaProducerConfig = checkNotNull(kafkaProducerConfig, "kafkaProducerConfig");
+        this.transactionalIdPrefix = checkNotNull(transactionalIdPrefix, "transactionalIdPrefix");
+        this.recordSerializer = checkNotNull(recordSerializer, "recordSerializer");
+        try {
+            recordSerializer.open(schemaContext);
+        } catch (Exception e) {
+            throw new RuntimeException("Cannot initialize schema.", e);
+        }
+        this.deliveryCallback =
+                (metadata, exception) -> {
+                    if (exception != null && producerAsyncException == null) {
+                        producerAsyncException = exception;
+                    }
+                    acknowledgeMessage();
+                };
+        this.metricGroup = sinkInitContext.metricGroup();
+        this.kafkaSinkContext =
+                new DefaultKafkaSinkContext(
+                        checkNotNull(sinkInitContext, "sinkInitContext"), kafkaProducerConfig);
+        this.kafkaWriterState =
+                recoverAndInitializeState(checkNotNull(recoveredStates, "recoveredStates"));
+        this.currentProducer = beginTransaction();
+        producers.add(currentProducer);
+    }
+
+    @Override
+    public void write(IN element, Context context) throws IOException {
+        checkErroneous();
+        final ProducerRecord<byte[], byte[]> record =
+                recordSerializer.serialize(element, kafkaSinkContext, context.timestamp());
+        pendingRecords.incrementAndGet();
+        currentProducer.send(record, deliveryCallback);
+    }
+
+    @Override
+    public List<KafkaCommittable> prepareCommit(boolean flush) throws IOException {
+        flushRecords(flush);
+        if (!flush) {
+            currentProducer = beginTransaction();
+        }
+        final List<KafkaCommittable> committables = commit();
+        producers.add(currentProducer);
+        return committables;
+    }
+
+    @Override
+    public List<KafkaWriterState> snapshotState() throws IOException {
+        return ImmutableList.of(kafkaWriterState);
+    }
+
+    @Override
+    public void close() throws Exception {
+        currentProducer.close(Duration.ZERO);
```

Comment:
producers for each?

Reply:
Every time a new producer is created we immediately close the old thus we do not need to do it here.